crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -1014,89 +1023,130 @@ struct RangeBlockMode<
     iter: Peekable<std::vec::IntoIter<SealedHeader<<Eth::Provider as HeaderProvider>::Header>>>,
     next: VecDeque<ReceiptBlockResult<Eth::Provider>>,
     max_range: usize,
+    // Stream of ongoing receipt fetching tasks
+    pending_tasks: FuturesOrdered<ReceiptFetchFuture<Eth::Provider>>,
 }
 
 impl<
         Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
     > RangeBlockMode<Eth>
 {
     async fn next(&mut self) -> Result<Option<ReceiptBlockResult<Eth::Provider>>, EthFilterError> {
+        // First, try to return any already processed results
         if let Some(result) = self.next.pop_front() {
             return Ok(Some(result));
         }
 
-        let Some(next_header) = self.iter.next() else {
-            return Ok(None);
-        };
+        // Next, try to get a completed task result if there are pending tasks
+        if !self.pending_tasks.is_empty() {
+            if let Some(task_result) = self.pending_tasks.next().await {
+                match task_result {
+                    Ok(chunk_results) => {
+                        for result in chunk_results {
+                            self.next.push_back(result);
+                        }
+                        return Ok(self.next.pop_front());
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+        }
+
+        // If no pending tasks and no queued results, try to start new tasks
+        if self.pending_tasks.is_empty() {
+            // Try to get the next header(s) to process
```

Comment:
here is an example of why this would be better as a proper `Stream` impl instead of custom method, because the appropriate thing to do here is schedule for wake up and return `Poll::Pending`, out of scope of this pr though. cc @mattsse @fgimenez